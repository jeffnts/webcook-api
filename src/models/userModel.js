import mongoose from 'mongoose'
const Schema = mongoose.Schema
import bcrypt from 'bcryptjs'

const userSchema = new Schema({

    name:{
        type: String,
        required: true
    },
    userName:{
        type: String,
        required: true,
        unique: true,
        lowercase: true,
        trim: true
    },
    email : {
        type: String,
        unique: true,
        required: true,
        lowercase: true,
        validate: function(email) {
            return /^[a-zA-Z0-9.!#$%&â€™*+\/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(email)
        }

    },
    password:{
        type: String,
        required: true,
        select: false
    },
    role:{
        type: String,
        required: true,
        enum:[
            'user',
            'admin'
        ],
        default: 'user'
    },
    photo:{
        type: String,
        default: 'avatar'
    },
    recipes: [
        {   type: Schema.Types.ObjectId,
            ref: 'Recipe'

        }

    ]


});
// Auth settings with Brcypt
    userSchema.pre('save', function hashPassword(next) {
        if (this.isModified('password')) {
            this.encryptPassword(this.password)
                .then(hash => {
                    this.password = hash;
                    next();
                })
                .catch(err => next(err));
        } else {
            return next();
        }
    });

    userSchema.methods = {
        async authenticate(plainText) {
            try {
                return await bcrypt.compare(plainText, this.password);
            } catch (err) {
                return false;
            }
        },
        encryptPassword(password) {
            return bcrypt.hash(password, 8);
        },
    };


export default  mongoose.model('User', userSchema);