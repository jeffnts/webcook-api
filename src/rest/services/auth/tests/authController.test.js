const app = require('../../../../app')
const request = require('supertest')(app.callback())
import userModel from '../../../../models/userModel'
const connection = require('../../../../config/database/database_tests')
const {sanitizeTestObject} = require('../../../../config/tests/sanitizeTestObject')

const dropUserCollection = async () => userModel.remove();


let user = {
    '_id': 'TESTOBJECT01',
    'name': 'Jhon',
    'email': 'jhon@email.com',
    'userName': 'jhon',
    'password': '123456'
};

const createUser = async () => userModel.create(user);

beforeEach(() => {
    createUser()
});

afterEach(()=> {
    dropUserCollection()
});

beforeAll(() => {
    dropUserCollection()
});

afterAll(() => {
    connection.disconect()
});

describe('Auth tests', ()=>{

    test('It should authenticate the user when it is passed the correct userName and password',async () => {

        const res = await request.post('/api/auth/login').send({'userName': user.userName, 'password': user.password})
        expect(sanitizeTestObject(res.body, ['token'])).toMatchSnapshot()
    });

    test('It should not authenticate the user when it is passed the wrong userName ',async () => {

        const res = await request.post('/api/auth/login').send({'userName': 'JÃµao', 'password': user.password})
        expect(sanitizeTestObject(res.body, ['token'])).toMatchSnapshot()
    });

    test('It should not authenticate the user when it is passed the wrong password',async () => {

        const res = await request.post('/api/auth/login').send({'userName': user.userName, 'password': '1'})
        expect(sanitizeTestObject(res.body, ['token'])).toMatchSnapshot()
    });

});