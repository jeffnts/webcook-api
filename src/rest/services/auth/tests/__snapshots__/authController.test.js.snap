// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth tests It should authenticate the user when it is passed the correct userName and password 1`] = `Object {}`;

exports[`Auth tests It should authenticate the user when it is passed the wrong password 1`] = `Object {}`;

exports[`Auth tests It should authenticate the user when it is passed the wrong userName  1`] = `Object {}`;
